File: app/src/main/java/com/mgaetan89/showsrage/fragment/ScheduleSectionFragment.java
package com.mgaetan89.showsrage.fragment;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.mgaetan89.showsrage.Constants;
import com.mgaetan89.showsrage.R;
import com.mgaetan89.showsrage.adapter.ScheduleAdapter;
import com.mgaetan89.showsrage.model.Schedule;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

public class ScheduleSectionFragment extends Fragment {
	@Nullable
	private ScheduleAdapter adapter = null;

	@Nullable
	private TextView emptyView = null;

	@Nullable
	private RecyclerView recyclerView = null;

	@NonNull
	private final List<Schedule> schedules = new ArrayList<>();

	public ScheduleSectionFragment() {
	}

	@Override
	public void onActivityCreated(Bundle savedInstanceState) {
		super.onActivityCreated(savedInstanceState);

		Bundle arguments = this.getArguments();

		if (arguments != null) {
			Collection<Schedule> schedules = (Collection<Schedule>) arguments.getSerializable(Constants.Bundle.INSTANCE.getSCHEDULES());

			if (schedules != null) {
				this.schedules.addAll(schedules);
			}
		}

		if (this.schedules.isEmpty()) {
			if (this.emptyView != null) {
				this.emptyView.setVisibility(View.VISIBLE);
			}

			if (this.recyclerView != null) {
				this.recyclerView.setVisibility(View.GONE);
			}
		} else {
			if (this.emptyView != null) {
				this.emptyView.setVisibility(View.GONE);
			}

			if (this.recyclerView != null) {
				this.recyclerView.setVisibility(View.VISIBLE);
			}
		}

		if (this.adapter != null) {
			this.adapter.notifyDataSetChanged();
		}
	}

	@Nullable
	@Override
	public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
		View view = inflater.inflate(R.layout.fragment_schedule_section, container, false);

		if (view != null) {
			this.emptyView = (TextView) view.findViewById(android.R.id.empty);
			this.recyclerView = (RecyclerView) view.findViewById(android.R.id.list);

			if (this.recyclerView != null) {
				int columnCount = this.getResources().getInteger(R.integer.shows_column_count);
				this.adapter = new ScheduleAdapter(this.schedules);

				this.recyclerView.setAdapter(this.adapter);
				this.recyclerView.setLayoutManager(new GridLayoutManager(this.getActivity(), columnCount));
			}
		}

		return view;
	}

	@Override
	public void onDestroy() {
		this.schedules.clear();

		super.onDestroy();
	}

	@Override
	public void onDestroyView() {
		this.emptyView = null;
		this.recyclerView = null;

		super.onDestroyView();
	}
}

File: app/src/main/kotlin/com/mgaetan89/showsrage/fragment/ScheduleSectionFragment.kt
package com.mgaetan89.showsrage.fragment

import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.GridLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.mgaetan89.showsrage.Constants
import com.mgaetan89.showsrage.R
import com.mgaetan89.showsrage.adapter.ScheduleAdapter
import com.mgaetan89.showsrage.extension.getSchedule
import com.mgaetan89.showsrage.model.Schedule
import io.realm.Realm
import io.realm.RealmChangeListener
import io.realm.RealmResults
import kotlinx.android.synthetic.main.fragment_schedule_section.empty
import kotlinx.android.synthetic.main.fragment_schedule_section.list

class ScheduleSectionFragment : Fragment(), RealmChangeListener<RealmResults<Schedule>> {
	private lateinit var realm: Realm
	private lateinit var schedules: RealmResults<Schedule>

	override fun onChange(schedules: RealmResults<Schedule>) {
		if (this.schedules.isEmpty()) {
			this.empty?.visibility = View.VISIBLE
			this.list?.visibility = View.GONE
		} else {
			this.empty?.visibility = View.GONE
			this.list?.visibility = View.VISIBLE
		}

		this.list?.adapter?.notifyDataSetChanged()
	}

	override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View?
			= inflater.inflate(R.layout.fragment_schedule_section, container, false)

	override fun onStart() {
		super.onStart()

		this.realm = Realm.getDefaultInstance()
		this.schedules = this.realm.getSchedule(this.arguments?.getString(Constants.Bundle.SCHEDULE_SECTION, "") ?: "", this)
		this.list?.adapter = ScheduleAdapter(this.schedules)
	}

	override fun onStop() {
		if (this.schedules.isValid) {
			this.schedules.removeAllChangeListeners()
		}

		this.realm.close()

		super.onStop()
	}

	override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
		val columnCount = this.resources.getInteger(R.integer.shows_column_count)

		this.list?.layoutManager = GridLayoutManager(this.activity, columnCount)
	}

	companion object {
		fun newInstance(section: String) = ScheduleSectionFragment().apply {
			this.arguments = Bundle().apply {
				this.putString(Constants.Bundle.SCHEDULE_SECTION, section)
			}
		}
	}
}
